edit-post/components/browser-url/index.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { withSelect } from '@wordpress/data';
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Returns the Post's Edit URL.
+ *
+ * @param {number} postId Post ID.
+ *
+ * @return {string} Post edit URL.
+ */
+export function getPostEditURL( postId ) {
+	return addQueryArgs( 'post.php', { post: postId, action: 'edit' } );
+}
+
+export class BrowserURL extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			historyId: null,
+		};
+	}
+
+	componentDidUpdate( prevProps ) {
+		const { postId, postStatus } = this.props;
+		const { historyId } = this.state;
+		if ( postId === prevProps.postId && postId === historyId ) {
+			return;
+		}
+
+		if ( postStatus !== 'auto-draft' ) {
+			this.setBrowserURL( postId );
+		}
+	}
+
+	/**
+	 * Replaces the browser URL with a post editor link for the given post ID.
+	 *
+	 * Note it is important that, since this function may be called when the
+	 * editor first loads, the result generated `getPostEditURL` matches that
+	 * produced by the server. Otherwise, the URL will change unexpectedly.
+	 *
+	 * @param {number} postId Post ID for which to generate post editor URL.
+	 */
+	setBrowserURL( postId ) {
+		window.history.replaceState(
+			{ id: postId },
+			'Post ' + postId,
+			getPostEditURL( postId )
+		);
+
+		this.setState( () => ( {
+			historyId: postId,
+		} ) );
+	}
+
+	render() {
+		return null;
+	}
+}
+
+export default withSelect( ( select ) => {
+	const { getCurrentPost } = select( 'core/editor' );
+	const { id, status } = getCurrentPost();
```

Comment:
I know that the current post is updated when a request triggers (and not when it succeed). Could this be an issue if the request fails and the status is reverted for instance?

Reply:
>Could this be an issue if the request fails and the status is reverted for instance?

Hmm, I suppose this could lead to some oddity where the UI is _too_ optimistic. It might be reproduceable for a new post where the status is explicitly set to something. When hitting save, I expect the URL will change immediately. If the request then fails, the `status` would be reverted back to `auto-draft` but, since we don't allow `replaceState` to be called for auto-draft, it will linger in its optimistic state. 

Might be worth discussing more what would actually be expected here. But, for what it's worth, in my brief testing I find that navigating to `post.php?id=XYZ` where XYZ is the ID of an auto-draft post, the editor loads just fine (i.e. no effective breakage).

Reply:
Yes, I'm not too concerned here as well. I guess the alternative would be to revert the url to `post-new` if the status is reverted to `auto-draft`

Reply:
>Yes, I'm not too concerned here as well. I guess the alternative would be to revert the url to `post-new` if the status is reverted to `auto-draft`

Yep, sounds reasonable. Since it's not breaking, I'll create an issue and we can address this after release.